bpf: introduce libbpf dependency; replace some bpf syscall with libbpf API

wat-ze-hex opened the pull request:
An attempt to replace direct bpf syscalls with libbpf was made in #12151 where libbpf was introduced as a git submodule.
This is another approach with libbpf as a library.
`libbpf` package build from GH mirror is available in Fedora [1]. 
Debian package exists as well but it is build from kernel sources [2]. 
There is no CentOS package but it's possible to make one by rebuild the Fedora package and publishing it to EPEL repo.
I expect initial test failures, because they may lack of libbpf package installed thus libbpf dependency unresolved.

[1] https://rpms.remirepo.net/rpmphp/zoom.php?rpm=libbpf
[2] https://packages.debian.org/sid/libbpf-dev

wat-ze-hex commented:
In order to push this forward, some help from the community is needed, so I'm tagging people :)
List of blockers:
1. Testing ecosystem need to support `libbpf` dependency. There is no CentOS package yet but publishing it to Fedora EPEL seems to be a promising approach. cc @evverx 
Evgeny, how hard it's to modify CentOS CI tests after `libbpf` is available there?
2. Secondary, we have Debian built from the kernel sources. We need to switch it to GH mirror. cc @rfc1036 
Marco, I wonder how hard it to file a bugreport similar to what' made in Fedora?[[1] ](https://lore.kernel.org/netdev/A273A3DB-C63E-488F-BB0C-B7B2AB6D99BE@fb.com/)
3. Absence of `libbpf` in Ubuntu
4. After a package is available in Debian and Ubuntu, libbpf should be made systemd dependency,  so it can be installed by `sudo apt-get build-dep systemd -y` command. This should help with `systemd.systemd` test.
5. There are also tests dependent on [[2]](https://salsa.debian.org/systemd-team/systemd): ubuntu bionic tests and SemaphoreCI  cc @martinpitt

So, a general question: let's pretend libbpf package is available, what are the steps to introduce it as a dependency to systemd and testing infra? How new package dependencies are usually added? cc @keszybz @mrc0mmand @poettering 

As an alternative, I don't mind reviving submodule approach #12151 but it also needs support from testing infra so ubuntu bionic tests are unblocked.

Thanks for all you replies!

[1] https://lore.kernel.org/netdev/A273A3DB-C63E-488F-BB0C-B7B2AB6D99BE@fb.com/
[2] https://salsa.debian.org/systemd-team/systemd

evverx commented:
@wat-ze-hex regarding CentOS CI, I believe `libbpf` can be added to https://github.com/systemd/systemd-centos-ci/blob/master/agent/bootstrap.sh#L45-L50 (where as far as I know, all the dependencies are installed). Generally @mrc0mmand maintains those scripts and knows what to do much better than me.

Regarding `systemd.systemd`, there the fuzzers are built and we generally skip all the dependencies so, by analogy with `libseccomp`, something like
```
want_libbpf = get_option('libbpf')
if want_libbpf != 'false' and not skip_deps
   ...
```
should make that part green.

evverx commented:
I've just realized that `libbpf` isn't supposed to be optional so skipping it when the fuzzers are built doesn't look right. To some extent we kind of get away with it (with `libmount` and `libcap`) because those libraries are also built with all the sanitizers from scratch on the OSS-Fuzz side but I'm not sure how hard it will be to bring `libbpf` there. Ideally we should build dependencies like that along with `systemd` and link them statically (which will automatically resolve https://github.com/systemd/systemd/pull/11730 and https://github.com/systemd/systemd/pull/12608) but nobody has got round to it yet. Let's just ignore `systemd.systemd` for now.

wat-ze-hex commented:
@evverx Just to clarify, do you mean `fuzzit.sh` script triggered by TravisCI and triggering `test/fuzz/test-.*` it its turn? My thought it is failing because libbpf package doesn't present in Ubuntu.
Yes, it's impossible to make libbpf optional.

evverx commented:
@wat-ze-hex I was talking about `fuzzbuzz.sh`, which among other things builds the fuzzers we run on OSS-Fuzz.

`fuzzit.sh` compiles the fuzzers as well and send them to Fuzzit, which we use for regression testing. It's true currently it's failing because it can't find `libbpf` on Ubuntu but even if `libbpf` was on Travis CI the fuzzers would fail because Travis CI is just a build environment and whatever we install there isn't available in the runtime environment. The same is applicable to OSS-Fuzz where the build environment (where it will be relatively easy to install `libbpf` once it's brought to Ubuntu) is different from the runtime environment (where build dependencies usually aren't available).

evverx commented:
@wat-ze-hex I think what I was trying to say is documented in https://github.com/google/oss-fuzz/blob/97060c44dec9c25b8c74a4ee05a217bd68f119fe/docs/further-reading/fuzzer_environment.md#runtime-dependencies.

When the fuzzers were introduced we kind of ignored that, made some assumptions about the runtime environment and somehow convinced the OSS-Fuzz team to add an undocumented feature allowing us to also use `libshared-systemd` linked dynamically. Generally it was a stopgap to get the fuzzers up and running and the whole thing can actually break any moment. Everybody else links their dependencies statically.

Ideally, we should fix the fuzzers but if it takes too long another option would be to use my YOLO approach from https://github.com/systemd/systemd/pull/12608 where both `libcap` and `libmount` were skipped when the fuzzers were built (which is, well, more or less OK because for the fuzzers those dependencies are technically optional).

Anyway, I don't think at this point it's important enough. I'm pretty sure we'll figure something out eventually. My understanding is that it would be better to decide where the PR is headed in general. I hope it makes sense.

keszybz commented on src/shared/bpf-program.c:
```diff
@@ -193,64 +170,12 @@ int bpf_program_cgroup_detach(BPFProgram *p) {
-                if (bpf(BPF_PROG_DETACH, &attr, sizeof(attr)) < 0)
+                if (bpf_prog_detach2(p->kernel_fd, fd, p->attached_type) < 0) {
                         return -errno;
+                }
```
Here too.

wat-ze-hex commented:
@evverx 
> @wat-ze-hex I think what I was trying to say is documented in https://github.com/google/oss-fuzz/blob/97060c44dec9c25b8c74a4ee05a217bd68f119fe/docs/further-reading/fuzzer_environment.md#runtime-dependencies.
> 
I see that runtime dependency is a concern. What about libbpf linked statically, i.e.:
```
libbpf = dependency('libbpf', required : false, static : true)
```
In the case of static linking there is still a question of how to link the dependencies of libbpf – `libelf` and `zlib` in their turn. If we can rely on their presence in runtime environments, they can be link dynamically. If we can't, we need to link `libz.a` and `libelf.a`. 
Wdyt?

wat-ze-hex commented:
> @wat-ze-hex as for the CentOS 7 job, I could rebuild the `libbpf` package and import it into _our_ systemd CentOS CI [copr repo](https://copr.fedorainfracloud.org/coprs/mrc0mmand/systemd-centos-ci/) until it's available in EPEL (if it's going to be available there). Not sure about the Arch Linux part of CentOS CI, though, as I have absolutely zero experience building packages there.

That would be very helpful, thank you! As for EPEL packaging it's WIP, the conversation can be tracked in [[1]](https://lore.kernel.org/netdev/20191008073958.GA10009@krava/)
Fedora package is in [[2]](https://apps.fedoraproject.org/packages/libbpf)


[1] https://lore.kernel.org/netdev/20191008073958.GA10009@krava/
[2] https://apps.fedoraproject.org/packages/libbpf

mbiebl commented:
fwiw, I think it's way too early to add a dependency on libbpf.
This should only happen once libbpf has seen more widespread usage and is available in (most) distros, ideally LTS distros.

wat-ze-hex commented on meson.build:
```diff
@@ -834,6 +834,16 @@ if not libcap.found()
+endif
+
+if not libbpf.found()
+        libbpf = cc.find_library('bpf')
```
@keszybz Thanks for review.
I've added comments:
`libbpf` is presently provided by pkg-config in Fedora and will be provided in CentOS.
`libbpf-dev` is the present Debian package built from kernel sources.
Removed redundant `find_library` branch.

evverx commented:
> If we can rely on their presence in runtime environments, they can be linked dynamically. If we can't, we need to link libz.a and libelf.a

I'll have to find out whether these are available there. Though, I was told (repeatedly) not to rely on anything. What I'm concerned about is that even if they are there it's going to be another stopgap we'll never fix (or, more precisely, the part linking `libbpf` will be correct but the rest will be the way it is) :-)

> This should only happen once libbpf has seen more widespread usage

@mbiebl Given that some (probably most) projects heavily relying on BPF (most notably https://github.com/iovisor/bcc) have already switched to `libbpf` (as a submodule usually) I'd say the library is as widespread as it can possibly be (considering the nature of the library). If that's not enough, I don't know what is. Plus, as far as I can tell, it's already tested automatically better than a lot of other open source projects that have seen widespread usage and nevertheless are failing to compile with `-Wall -Wextra`. (`systemd` also fails to compile with `-Wall -Wextra` but that doesn't count obviously :-))

boucman commented:
Well, I did a quick search and bcc seems to be supported neither in yocto nor buildroot afaict (there are some patches under buildroot, but they are still under consideration). I couldn't find libbpf either, but I didn't dig that deep.

This is usually a strong sign that those libraries do not support cross-compilation which is essential for the embedded world. So... this would probably block any systemd upgrade in the embedded world, which is bad. especially if it's a mandatory dependency.

evverx commented:
> This is usually a strong sign that those libraries do not support cross-compilation which is essential for the embedded world.

@boucman have you tried to cross-compile it? If it doesn't work, it's possible to open an issue on their bug tracker at https://github.com/libbpf/libbpf. All issues I opened there were closed in less than a week so in my experience the project is very responsive. I'd also like to point out that `systemd` itself, as you might have noticed, have some issues with cross-compilation from time to time and relies on bug reports and patches from people who care about it.

evverx commented:
> especially if it's a mandatory dependency.

I'm not sure if it was discussed at All Systems Go but given that it's going to be a normal dependency (as opposed to a manually implemented part of systemd), it seems it would make sense to make it possible to make it optional like almost everything else. Another option would be to keep using the old stuff if the library isn't available (so that the BPF part of systemd won't stop working suddenly) effectively cementing the status quo where libbpf won't be seen in the foreseeable future but without any updates and bug fixes. New features and bug fixes would be present in the maintained "libbpf" part only and if those features and bug fixes are important enough for, for example, embedded folks they would find to a way to bring libbpf there.

wat-ze-hex commented:
@mbiebl  

> fwiw, I think it's way too early to add a dependency on libbpf.
> This should only happen once libbpf has seen more widespread usage and is available in (most) distros, ideally LTS distros.

That's a chicken and egg problem: packages don't pop up by themselves unless driven by a use case, a use case is blocked by an absence of packages. `systemd` is a perfect example of a use case because the present systemd BPF infra... well... sucks.
So we can wait until someone do a dirty job for us, or we can drive it. The problem with _wait_ is that ETA is unpredictable and in the worst case infinite.
I'm flexible and can go back to submodule approach #12151 as a compromise.

mrc0mmand commented:
> > @wat-ze-hex as for the CentOS 7 job, I could rebuild the `libbpf` package and import it into _our_ systemd CentOS CI [copr repo](https://copr.fedorainfracloud.org/coprs/mrc0mmand/systemd-centos-ci/) until it's available in EPEL (if it's going to be available there). Not sure about the Arch Linux part of CentOS CI, though, as I have absolutely zero experience building packages there.
> 
> That would be very helpful, thank you! As for EPEL packaging it's WIP, the conversation can be tracked in [[1]](https://lore.kernel.org/netdev/20191008073958.GA10009@krava/)
> Fedora package is in [[2]](https://apps.fedoraproject.org/packages/libbpf)
> 
> [1] https://lore.kernel.org/netdev/20191008073958.GA10009@krava/
> [2] https://apps.fedoraproject.org/packages/libbpf

The COPR EPEL7 build is [pending](https://copr.fedorainfracloud.org/coprs/mrc0mmand/systemd-centos-ci/build/1053694/), shouldn't take much longer. `fedpkg --relese epel7 mockbuild` passed without an issue, so it should be a matter of simply adding a new dependency in the CentOS 7 job once the package is built.

As for the Arch Linux, I guess I'll finally have to learn how to use AUR, so we can test it there as well.

mrc0mmand commented:
@wat-ze-hex the CentOS 7 job finally got past the meson setup phase:

```
Run-time dependency libbpf found: YES 0.0.5
```

mrc0mmand commented:
@wat-ze-hex as for Arch Linux - I wrote [a simple](https://gist.github.com/mrc0mmand/b004e519d46eeb9fe669f63f1f33730f) `PKGBUILD` file (inspired by the respective Fedora SPEC) and it seems to be working as intended:

```
# sudo -u builder makepkg
==> Making package: libbpf 0.0.5-1 (Fri 11 Oct 2019 08:38:18 PM UTC)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Downloading v0.0.5.tar.gz...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   121    0   121    0     0    384      0 --:--:-- --:--:-- --:--:--   384
100  156k    0  156k    0     0   139k      0 --:--:--  0:00:01 --:--:--  200k
==> Validating source files with md5sums...
    v0.0.5.tar.gz ... Skipped
==> Extracting sources...
  -> Extracting v0.0.5.tar.gz with bsdtar
==> Starting build()...
mkdir -p ./staticobjs
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c bpf.c -o staticobjs/bpf.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c btf.c -o staticobjs/btf.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c libbpf.c -o staticobjs/libbpf.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c libbpf_errno.c -o staticobjs/libbpf_errno.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c netlink.c -o staticobjs/netlink.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c nlattr.c -o staticobjs/nlattr.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c str_error.c -o staticobjs/str_error.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c libbpf_probes.c -o staticobjs/libbpf_probes.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c bpf_prog_linfo.c -o staticobjs/bpf_prog_linfo.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c xsk.c -o staticobjs/xsk.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c btf_dump.c -o staticobjs/btf_dump.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_FORTIFY_SOURCE=2 -c hashmap.c -o staticobjs/hashmap.o
ar rcs libbpf.a staticobjs/bpf.o staticobjs/btf.o staticobjs/libbpf.o staticobjs/libbpf_errno.o staticobjs/netlink.o staticobjs/nlattr.o staticobjs/str_error.o staticobjs/libbpf_probes.o staticobjs/bpf_prog_linfo.o staticobjs/xsk.o staticobjs/btf_dump.o staticobjs/hashmap.o
mkdir -p ./sharedobjs
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c bpf.c -o sharedobjs/bpf.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c btf.c -o sharedobjs/btf.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c libbpf.c -o sharedobjs/libbpf.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c libbpf_errno.c -o sharedobjs/libbpf_errno.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c netlink.c -o sharedobjs/netlink.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c nlattr.c -o sharedobjs/nlattr.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c str_error.c -o sharedobjs/str_error.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c libbpf_probes.c -o sharedobjs/libbpf_probes.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c bpf_prog_linfo.c -o sharedobjs/bpf_prog_linfo.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c xsk.c -o sharedobjs/xsk.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c btf_dump.c -o sharedobjs/btf_dump.o
cc -I. -I../include -I../include/uapi -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -fvisibility=hidden -DSHARED -D_FORTIFY_SOURCE=2 -c hashmap.c -o sharedobjs/hashmap.o
cc -shared -Wl,--version-script=libbpf.map \
	      -Wl,-soname,libbpf.so.0 \
	      sharedobjs/bpf.o sharedobjs/btf.o sharedobjs/libbpf.o sharedobjs/libbpf_errno.o sharedobjs/netlink.o sharedobjs/nlattr.o sharedobjs/str_error.o sharedobjs/libbpf_probes.o sharedobjs/bpf_prog_linfo.o sharedobjs/xsk.o sharedobjs/btf_dump.o sharedobjs/hashmap.o -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -lelf -o libbpf.so.0.0.5
ln -sf libbpf.so.0.0.5 libbpf.so.0
ln -sf libbpf.so.0 libbpf.so
sed -e "s|@PREFIX@|/usr|" \
	-e "s|@LIBDIR@|/usr/lib64|" \
	-e "s|@VERSION@|0.0.5|" \
	< libbpf.pc.template > libbpf.pc
==> Entering fakeroot environment...
==> Starting package()...
if [ ! -d '/workspace/pkg/libbpf//usr/include/bpf' ]; then install -d -m 755 '/workspace/pkg/libbpf//usr/include/bpf'; fi; install bpf.h libbpf.h btf.h xsk.h libbpf_util.h -m 644 '/workspace/pkg/libbpf//usr/include/bpf'
if [ ! -d '/workspace/pkg/libbpf//usr/lib64/pkgconfig' ]; then install -d -m 755 '/workspace/pkg/libbpf//usr/lib64/pkgconfig'; fi; install ./libbpf.pc -m 644 '/workspace/pkg/libbpf//usr/lib64/pkgconfig'
if [ ! -d '/workspace/pkg/libbpf//usr/lib64' ]; then install -d -m 755 '/workspace/pkg/libbpf//usr/lib64'; fi; cp -fpR ./libbpf.a ./libbpf.so ./libbpf.so.0 ./libbpf.so.0.0.5 '/workspace/pkg/libbpf//usr/lib64'
==> Tidying install...
  -> Removing libtool files...
  -> Purging unwanted files...
  -> Stripping unneeded symbols from binaries and libraries...
  -> Compressing man and info pages...
==> Checking for packaging issues...
==> Creating package "libbpf"...
  -> Generating .PKGINFO file...
  -> Generating .BUILDINFO file...
  -> Generating .MTREE file...
  -> Compressing package...
==> Leaving fakeroot environment.
==> Finished making: libbpf 0.0.5-1 (Fri 11 Oct 2019 08:38:28 PM UTC)
```

```
# tar tvf libbpf-0.0.5-1-x86_64.pkg.tar.xz 
-rw-r--r-- root/root       403 2019-10-11 20:38 .PKGINFO
-rw-r--r-- root/root      5030 2019-10-11 20:38 .BUILDINFO
-rw-r--r-- root/root       930 2019-10-11 20:38 .MTREE
drwxr-xr-x root/root         0 2019-10-11 20:38 usr/
drwxr-xr-x root/root         0 2019-10-11 20:38 usr/include/
drwxr-xr-x root/root         0 2019-10-11 20:38 usr/include/bpf/
-rw-r--r-- root/root      6416 2019-10-11 20:38 usr/include/bpf/bpf.h
-rw-r--r-- root/root      8195 2019-10-11 20:38 usr/include/bpf/btf.h
-rw-r--r-- root/root     19867 2019-10-11 20:38 usr/include/bpf/libbpf.h
-rw-r--r-- root/root      1643 2019-10-11 20:38 usr/include/bpf/libbpf_util.h
-rw-r--r-- root/root      6289 2019-10-11 20:38 usr/include/bpf/xsk.h
drwxr-xr-x root/root         0 2019-10-11 20:38 usr/lib64/
-rw-r--r-- root/root    226596 2019-10-11 20:38 usr/lib64/libbpf.a
lrwxrwxrwx root/root         0 2019-10-11 20:38 usr/lib64/libbpf.so -> libbpf.so.0
lrwxrwxrwx root/root         0 2019-10-11 20:38 usr/lib64/libbpf.so.0 -> libbpf.so.0.0.5
-rwxr-xr-x root/root    141216 2019-10-11 20:38 usr/lib64/libbpf.so.0.0.5
drwxr-xr-x root/root         0 2019-10-11 20:38 usr/lib64/pkgconfig/
-rw-r--r-- root/root       241 2019-10-11 20:38 usr/lib64/pkgconfig/libbpf.pc
```


Is it worth pushing into the Arch User Repository (AUR) or you already have plans to do it yourself (either to AUR or core)?

wat-ze-hex commented:
> Another option would be to keep using the old stuff if the library isn't available (so that the BPF part of systemd won't stop working suddenly) effectively cementing the status quo where libbpf won't be seen in the foreseeable future but without any updates and bug fixes.

1. This still means two things to maintain. No new features for old infra sound fine, but no maintenance at all (e.g. no bug fixes) for functionality which does something useful sounds unrealistic.
Minor note regarding no features: I have PR #13496 which, if merged, violates your proposal :)
2. A user will have to keep in mind that which e.g. unit file option invokes BPF infra v. 1 and which v. 2. That can be confusing.
3. In the ideal world the infra part may be static but same can't be applied for BPF programs – firewall and devices. Presence of both v. 1 and v. 2 implies that there will be two versions of programs which complicates things even more. Poor user.
4. Will there be End-of-Life for v. 1? It can't be there forever, isn't it?
5. Presence of the old stuff slows down evolvement of new stuff due to less feature request, bug reports and on-boarding experience.
6. Overhead for implementing, maintaining and deprecating v. 1 <-> v. 2 switch.

Summarizing, I would think about cementing the present stuff (which is maintenance anyway) only if there is no other way. And only with proclaimed EoL date.

wat-ze-hex commented:
@mrc0mmand 

Thank you! I mentioned that in the kernel mailing list.
As for Arch Linux, what's the procedure of maintenance? Once it's there at least `libbpf` version needs to be bumped along with the one in GH mirror. Would you like to do that? :) [[1]](https://help.github.com/en/articles/watching-and-unwatching-releases-for-a-repository) to track `libbpf` updates.

Alternatively, I can do that. Could you please provide a brief instruction due to I haven't published anything to AUR yet.

And thanks again.
[1] https://help.github.com/en/articles/watching-and-unwatching-releases-for-a-repository

wat-ze-hex commented:
@boucman 

> This is usually a strong sign that those libraries do not support cross-compilation which is essential for the embedded world. So... this would probably block any systemd upgrade in the embedded world, which is bad. especially if it's a mandatory dependency.

Could you please file a bug report to [[1]](https://github.com/libbpf/libbpf/issues) if you see any issues with cross compilation?

We can also think about a test for that, if given proper repro steps. Is there a test which checks x-compilation for `systemd`?

[1] https://github.com/libbpf/libbpf/issues

mrc0mmand commented:
> @mrc0mmand
> 
> Thank you! I mentioned that in the kernel mailing list.
> As for Arch Linux, what's the procedure of maintenance? Once it's there at least `libbpf` version needs to be bumped along with the one in GH mirror. Would you like to do that? :) [[1]](https://help.github.com/en/articles/watching-and-unwatching-releases-for-a-repository) to track `libbpf` updates.
> 
> Alternatively, I can do that. Could you please provide a brief instruction due to I haven't published anything to AUR yet.
> 
> And thanks again.
> [1] https://help.github.com/en/articles/watching-and-unwatching-releases-for-a-repository

I'll look into that, as I'll have to monitor the libbpf version anyway for the COPR repo (at least until its present in the EPEL) :-)

mrc0mmand commented:
@wat-ze-hex hmm, I noticed that libbpf seems to already part of this AUR package: https://aur.archlinux.org/packages/bcc-tools/, so I guess we could use that.

mrc0mmand commented:
Anyway, frankly, using AUR is not that much different than just cloning the [libbpf](https://github.com/libbpf/libbpf) repo and compiling it (as AUR packages have to be compiled locally anyway). Thus, I'm inclining to just adding the `git clone && make && make install` combo to the bootstrap phase and be done with it, as it sound much easier than dealing with the AUR machinery.

wat-ze-hex commented:
> @wat-ze-hex hmm, I noticed that libbpf seems to already part of this AUR package: https://aur.archlinux.org/packages/bcc-tools/, so I guess we could use that.

This looks fine with the exception that we don't need `bcc` portion of it. `libbpf` comes as a submodule of bcc and the version is bumped  with a [commit](https://github.com/iovisor/bcc/commit/eb018cf9e5df30c9748e4b31a33f0934a47cca6b).
There is a discussion to make `libbpf` a package [dependency](https://github.com/iovisor/bcc/pull/2535), so things may change.

So yeah, let's use use bcc in Arch tests.

mrc0mmand commented:
@wat-ze-hex could you give this one a force push? I decided to not use the `bcc` package, as it would complicate things a bit, see https://github.com/systemd/systemd-centos-ci/pull/175. But it should achieve the same thing, maybe even better, as the image is rebuilt every three days, thus the libbpf will be updated from the latest master as well.

keszybz commented on meson.build:
```diff
@@ -834,6 +834,15 @@ if not libcap.found()
+# Debian: libbpf packaged from kernel sources
+if not libbpf.found()
+        libbpf = dependency('libbpf-dev', required : true)
+endif
```
Hmm, I still don't get what libbpf-dev is here. `dependency` takes a pkg-config name as argument
(see https://mesonbuild.com/Reference-manual.html#dependency). Debian's `libbpf-dev` package has `libbpf.pc` file. I don't think anything except the first check is needed.

wat-ze-hex replied:
@keszybz 
You are right, checked that in my fork.

wat-ze-hex commented:
@keszybz

> I'm pretty sure we need to have the build dependency in Fedora, Centos, Debian, Ubuntu before we merge this. 

+1

> https://bugzilla.redhat.com/show_bug.cgi?id=1762219 is for Centos. I think the others are all covered. 

Thanks for filing this! Ubuntu is not covered. There is [bcc-tools](https://packages.ubuntu.com/bionic/all/libbpfcc-dev/filelist) but it doesn't provide libbpf like the one in Arch. 

> I would prefer to make this optional — i.e. simply skip relevant parts as not-available if the dependency is missing. I'm sure this will make the transition easier.

That would be ideal if it's OK to have BPF firewall and device disabled.

wat-ze-hex commented:
@mrc0mmand

> But it should achieve the same thing, maybe even better, as the image is rebuilt every three days, thus the libbpf will be updated from the latest master as well.


Thanks, CentOS CI part is green.

keszybz commented:
> Ubuntu is not covered.

Can you file a bug?

> That would be ideal if it's OK to have BPF firewall and device disabled.

I don't want to push a patch that will cause build failures. libbpf *is* rather new, and people will be unhappy. So if we want this to happen in a finite time, I think making the dep optional is the best best. It *will* be additional work, no doubt about that.

ssahani commented:
> I don't want to push a patch that will cause build failures. libbpf _is_ rather new, and people will be unhappy. So if we want this to happen in a finite time, I think making the dep optional is the best best. It _will_ be additional work, no doubt about that.

This will be a build failure for photon https://github.com/vmware/photon

michaelolbrich commented:
I did some cross-compile tests with libbpf. There are some minor issues, but nothing complicated. So no problems there.
But please make the dependency optional. For cross-compiling it's sometimes necessary to build a host systemd (e.g. for systemd-hwdb) and any unnecessary dependency hurts there.

wat-ze-hex commented:
@keszybz 
> > Ubuntu is not covered.
> 
> Can you file a bug?
> 

Since Ubuntu is Deb at first I'll file a bug to Deb to switch to GH mirror, then follow with Ubuntu with a request to use the Deb package.

> > That would be ideal if it's OK to have BPF firewall and device disabled.
> 
> I don't want to push a patch that will cause build failures. libbpf _is_ rather new, and people will be unhappy. So if we want this to happen in a finite time, I think making the dep optional is the best best. It _will_ be additional work, no doubt about that.

Do you mean the same that @evverx proposed which is cementing the current stuff or making the present BPF programs and infra optional if no package present?

wat-ze-hex commented:
@keszybz

Filed a bug for Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=942903

keszybz commented:
>> > Ubuntu is not covered.
>> 
>> Can you file a bug?
>
> https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=942903

There seems to have been a misunderstanding. I took your statement to mean that the library is not available. But it is available. It is probably better to switch to the standalone sources at some point, as far as building systemd is concerned, we don't really care how the library is built. So this bug is not a dependency for going through with this PR here.

> Do you mean the same that @evverx proposed which is cementing the current stuff or making the present BPF programs and infra optional if no package present?

I don't understand what @evverx is saying. I'm asking for the dependency to be optional. If the dependency is missing, systemd would be built without the functionality that depends on libbpf. This should work fine for fuzzing.

evverx commented:
>  Is there a test which checks x-compilation for systemd?

@wat-ze-hex no, there isn't. This part has never been tested upstream and is kept afloat thanks to bug reports and pull requests from people who care about it.

> I don't understand what @evverx is saying.

@keszybz what I was trying to say is that there are other options apart from making the dependency optional mostly to make sure that we know what we're doing by turning off the BPF stuff completely if the library isn't available. 

> If the dependency is missing, systemd would be built without the functionality that depends on libbpf.

Does that mean that it's OK to turn off the features (which have always been built-in and as far as I can remember advertised as related to security and sandboxing) to make it a little bit easier to build `systemd`?

evverx commented:
FWIW I've just taken a look at the documentation and it appears that it already says there that those features (like many other sandboxing features) should be used with the caveat that nobody should expect them to always work or rely on them for anything so technically, according to the documentation, in a way, it seems to be probably OK to make the dependency optional.

keszybz commented:
> @keszybz what I was trying to say is that there are other options apart from making the dependency optional mostly to make sure that we know what we're doing by turning off the BPF stuff completely if the library isn't available.

Which "other option" would you prefer?

> Does that mean that it's OK to turn off the features (which have always been built-in and as far as I can remember advertised as related to security and sandboxing) to make it a little bit easier to build systemd?

Yes. We can make the dependency explicitly opt-out to make sure it doesn't get turned off inadvertently.

evverx commented:
@keszybz in terms of phasing out the bpf stuff gradually, I'd prefer to use a variation of what seems to be a standard way to deal with things like mentioned in https://github.com/systemd/systemd/pull/13744#issuecomment-541130889:

> Another option would be to keep using the old stuff if the library isn't available (so that the BPF part of systemd won't stop working suddenly) effectively cementing the status quo where libbpf won't be seen in the foreseeable future but without any updates and bug fixes. New features and bug fixes would be present in the maintained "libbpf" part only and if those features and bug fixes are important enough for, for example, embedded folks they would find to a way to bring libbpf there.

but I agree with @wat-ze-hex that it's going to be a mess and judging by the comments in this thread nobody is particularly interested in phasing it out gradually. So I think it would make sense to make the dependency optional, wait for large distributions to roll it out, find the vulnerabilities that were dismissed because they were mitigated by the presence of the IP filters and file "security" issues against those distributions. I'm pretty sure this should help to package `libbpf` relatively fast :-)

evverx commented:
In the meantime, @mrc0mmand has set up arm builds on Travis CI to hopefully prevent issues like https://github.com/libbpf/libbpf/issues/83 from happening again and turned on LGTM for libbpf: https://lgtm.com/projects/g/libbpf/libbpf/alerts/?mode=list. Coverity Scan should be on the way I believe :-) Though, I've just updated the project there manually.

evverx commented:
FWIW @mrc0mmand integrated Coverity there as well (which went smoothly thanks to the Travis CI scripts @wat-ze-hex added back in April). All (!) the issues found by LGTM and Coverity were fixed so fast that I couldn't even take a closer look at what exactly those issues were about :-) As far as I can tell, it's the first library (I've ever seen) where there is nothing left for those static analyzers to complain about.

wat-ze-hex commented:
@evverx @keszybz 
Made BPF dependent part of systemd optional by introducing HAVE_LIBBPF macro and dummy implementations of BPF dependent functions. BPF dependency is deeply rooted hence used dummies instead of tons of `#if #else #endif` macro. Please verify if this approach looks good.

Some tests failed but all failures seem unrelated. cc @mrc0mmand 
CentOS CI (Arch in KVM): a transilient issues, probably will fix after rerun
bionic: all due to 
```
dpkg: error processing package systemd (--configure):
 installed systemd package post-installation script subprocess returned error exit status 1
```
semaphoreci: test-execute is broken on master:
```
GID 4 is no longer referenced, cleaning up its IPC.
Failed to connect to nscd socket: No such file or directory
Failed to connect to nscd socket: No such file or directory
	PID: 153730
	Start Timestamp: Sun 2019-12-01 20:42:10 PST
	Exit Timestamp: Sun 2019-12-01 20:42:10 PST
	Exit Code: exited
	Exit Status: 1
test_exec_dynamicuser: exec-dynamicuser-fixeduser-adm.service: exit status 1, expected 0
Aborted (core dumped)
```

evverx commented:
@wat-ze-hex I built `systemd` without `libbpf` and it appears `test-execute` is failing somewhere in `exec-privatedevices-yes.service`:
```sh
BPF is not supported. Mock is called
BPF is not supported. Mock is called
Unit exec-privatedevices-yes.service configures device ACL, but the local system doesn't seem to support the BPF-based device controller.
Proceeding WITHOUT applying ACL (all devices will be accessible)!
(This warning is only shown for the first loaded unit using device ACL.)
exec-privatedevices-yes.service: Passing 0 fds to service
exec-privatedevices-yes.service: About to execute: /bin/sh -c 'test ! -c /dev/kmsg'
exec-privatedevices-yes.service: Forked /bin/sh as 14740
exec-privatedevices-yes.service: Changed dead -> start
exec-privatedevices-yes.service
	UMask: 0022
	WorkingDirectory: /root
	RootDirectory: /
	NonBlocking: no
	PrivateTmp: no
	PrivateDevices: yes
	ProtectKernelTunables: no
	ProtectKernelModules: no
	ProtectKernelLogs: no
	ProtectControlGroups: no
	PrivateNetwork: no
	PrivateUsers: no
	ProtectHome: no
	ProtectSystem: no
	MountAPIVFS: no
	IgnoreSIGPIPE: yes
	MemoryDenyWriteExecute: no
	RestrictRealtime: no
	RestrictSUIDSGID: no
	KeyringMode: inherit
	ProtectHostname: no
	RuntimeDirectoryPreserve: no
	RuntimeDirectoryMode: 0755
	StateDirectoryMode: 0755
	CacheDirectoryMode: 0755
	LogsDirectoryMode: 0755
	ConfigurationDirectoryMode: 0755
	TimeoutCleanSec: infinity
	StandardInput: null
	StandardOutput: inherit
	StandardError: inherit
	CapabilityBoundingSet: cap_chown cap_dac_override cap_dac_read_search cap_fowner cap_fsetid cap_kill cap_setgid cap_setuid cap_setpcap cap_linux_immutable cap_net_bind_service cap_net_broadcast cap_net_admin cap_net_raw cap_ipc_lock cap_ipc_owner cap_sys_module cap_sys_chroot cap_sys_ptrace cap_sys_pacct cap_sys_admin cap_sys_boot cap_sys_nice cap_sys_resource cap_sys_time cap_sys_tty_config cap_lease cap_audit_write cap_audit_control cap_setfcap cap_mac_override cap_mac_admin cap_syslog cap_wake_alarm cap_block_suspend cap_audit_read
	DynamicUser: no
	LockPersonality: no
exec-privatedevices-yes.service: Failed to set up standard input: Operation not permitted
exec-privatedevices-yes.service: Failed at step STDIN spawning /bin/sh: Operation not permitted
Received SIGCHLD from PID 14740 ((sh)).
Child 14740 ((sh)) died (code=exited, status=208/STDIN)
exec-privatedevices-yes.service: Failed to read oom_kill field of memory.events cgroup attribute: No such file or directory
exec-privatedevices-yes.service: Child 14740 belongs to exec-privatedevices-yes.service.
exec-privatedevices-yes.service: Main process exited, code=exited, status=208/STDIN
exec-privatedevices-yes.service: Failed with result 'exit-code'.
exec-privatedevices-yes.service: Service will not restart (restart setting)
exec-privatedevices-yes.service: Changed start -> failed
exec-privatedevices-yes.service: Unit entered failed state.
	PID: 14740
	Start Timestamp: Mon 2019-12-02 17:38:57 UTC
	Exit Timestamp: Mon 2019-12-02 17:38:57 UTC
	Exit Code: exited
	Exit Status: 208
test_exec_privatedevices: exec-privatedevices-yes.service: exit status 208, expected 0
Aborted
```
Looks like this test should be skipped when `libbpf` isn't available. I suspect once it's done the test will no longer fail on Semaphore (where judging by "Dependency libbpf found: NO (tried pkgconfig and cmake)" `libbpf` isn't installed).

I'm not sure what's going on on Ubuntu CI (and I'm not going to launch an Ubuntu VM to just look at the logs) but it appears `systemd` is failing to restart `systemd-resolved`. It could be it has something to do with `PrivateDevices=true`. @ddstreet given that it's not the first we've had trouble figuring out what exactly goes wrong when the package is installed there I'm wondering if it would be possible to save the journal somewhere? For the time being, it would probably make sense for someone interested in keeping Ubuntu CI afloat to run the tests manually and paste the logs there.

evverx commented:
> I suspect once it's done the test will no longer fail on Semaphore

On the other hand, on Semaphore the tests are run in lxc containers and in theory `test_exec_privatedevices()` should be skipped there because it calls `detect_container()`. Looks like avoiding `test_exec_privatedevices` when `libbpf` isn't available won't help much there (but it's probably the right thing to do anyway to make the test pass on VMs, for example). I hope Semaphore won't cut the output of `test-execute` next time.

evverx commented:
I've just run the test in an lxc container manually and it failed there with:
```sh
BPF is not supported. Mock is called
BPF is not supported. Mock is called
Unit exec-readonlypaths.service configures device ACL, but the local system doesn't seem to support the BPF-based device controller.
Proceeding WITHOUT applying ACL (all devices will be accessible)!
(This warning is only shown for the first loaded unit using device ACL.)
exec-readonlypaths.service: Passing 0 fds to service
exec-readonlypaths.service: About to execute: /bin/sh -x -c 'test ! -w /usr && test ! -w /etc && test ! -w /sys && test ! -w /sys/fs/cgroup'
exec-readonlypaths.service: Forked /bin/sh as 31728
exec-readonlypaths.service: Changed dead -> start
exec-readonlypaths.service: Failed to set up standard input: Operation not permitted
exec-readonlypaths.service: Failed at step STDIN spawning /bin/sh: Operation not permitted
Received SIGCHLD from PID 31728 ((sh)).
Child 31728 ((sh)) died (code=exited, status=208/STDIN)
exec-readonlypaths.service: Child 31728 belongs to exec-readonlypaths.service.
exec-readonlypaths.service: Main process exited, code=exited, status=208/STDIN
exec-readonlypaths.service: Failed with result 'exit-code'.
exec-readonlypaths.service: Service will not restart (restart setting)
exec-readonlypaths.service: Changed start -> failed
exec-readonlypaths.service: Unit entered failed state.
test_exec_readonlypaths: exec-readonlypaths.service: exit status 208, expected 0
```
I'm not sure why it's complaining about STDIN though. As far as I can tell, something weird happens when `libbpf` isn't installed and a unit file has `PrivateDevices=yes`.

evverx commented:
I think I more or less understand what's going on. When `cg_all_unified()` <= 0, `cgroup_apply_devices` writes `a` to `devices.deny` and after that can't turn on anything because 
 all the `bpf_devices_*` functions don't work when libbpf isn't available. Looks like `cgroup_apply_devices` should be skipped altogether without `libbpf`. I haven't tried to run the tests with the unified hierarchy but it appears it should be functional there because when `cg_all_unified > 0` the function bails out as soon as `bpf_devices_cgroup_init` returns -1. Hope that makes sense :-)

evverx commented:
I'm not sure why Semaphore keeps truncating the build log (probably it has something to do with `autopkgtest` dumping absolutely everything when the package can't be built due to the unit tests (instead of showing just the output of failed tests)).

Judging by the number of jobs in the queues on Ubuntu CI (where currently the jobs submitted at  09:29:38 on 2019-12-02 are being run) it got stuck completely. @ddstreet I'm wondering if it's a glitch of some kind.
![Screenshot 2019-12-03 at 04 00 47](https://user-images.githubusercontent.com/10195800/70007492-8d12b780-1581-11ea-95d0-113c3c31ef94.png)


Anyway, it appears neither Semaphore nor Ubuntu CI is helpful at this point. To reproduce the issue locally it seems to be enough to build `systemd` without `libbpf` and run `test-execute` using the hybrid or legacy hierarchy where it should fail with something like "Main process exited, code=exited, status=208/STDIN".

wat-ze-hex commented:
Seems that the present `bionic` failures in systemd-nspawn are due to the devices previously whitelisted by BPF, e.g. `/dev/tty` are not whitelisted anymore:

```Failed to allocate a pty: Operation not permitted```

That needs a fix.

ddstreet commented:
> @ddstreet given that it's not the first we've had trouble figuring out what exactly goes wrong when the package is installed there I'm wondering if it would be possible to save the journal somewhere? 

the journal is saved for some tests, in the 'artifacts' tarball; but for test-exec (run by the 'root-unittests' autopkgtest) the journal doesn't get saved.  I'll open a Debian MR to update the test to save the journal on failure, but even then it will still be in the 'artifacts' tarball.  You can get that tarball by using the same URL for the log.gz file, but replace 'log.gz' at the end with 'artifacts.tar.gz'.  The Ubuntu CI server could use improvement since that's not obvious at all.

> Judging by the number of jobs in the queues on Ubuntu CI (where currently the jobs submitted at 09:29:38 on 2019-12-02 are being run) it got stuck completely. @ddstreet I'm wondering if it's a glitch of some kind.

I'm not exactly sure, but there were infrastructure problems earlier this week that might be the cause.

You can check this page: https://status.admin.canonical.com/
It doesn't specifically list status of the Ubuntu CI lab, but when it shows lots of stuff down (like on Dec 2) that probably means the Ubuntu CI test systems will have problems, too.

wat-ze-hex commented:
@ddstreet 

Any advice on how to repro `autopkgtest` failure?

Keeping running into this:
```timedated            PASS
hostnamed            PASS
localed-locale       PASS
localed-x11-keymap   PASS
logind               PASS
unit-config          PASS
storage              PASS
networkd-test.py     PASS
build-login          PASS
boot-and-services    FAIL non-zero exit status 1
udev                 PASS
root-unittests       PASS
upstream             FAIL non-zero exit status 1
boot-smoke           PASS
```
Failing w/ e.g. `Failed to allocate a pty: Operation not permitted`.
That is probably has something to do with the changes I've made. Thanks!

martinpitt commented:
See https://www.freedesktop.org/wiki/Software/systemd/autopkgtest/ for how to locally reproduce and investigate the test failures.

wat-ze-hex commented:
@keszybz @evverx 
Ok, figured it out. It's green. Review, please?

mrc0mmand commented:
@wat-ze-hex how safe is to drop the newly-introduced `kernel-headers >= 5.4.0-0.rc6.git0.1`[0] dep in the Fedora `libbpf` (`libbpf-devel`) package? I was about to bump the `libbpf` package in our CentOS 7 systemd repo[1], but this dependency would render it uninstallable.

[0] https://src.fedoraproject.org/rpms/libbpf/blob/master/f/libbpf.spec#_28
[1] https://copr.fedorainfracloud.org/coprs/mrc0mmand/systemd-centos-ci/package/libbpf/

wat-ze-hex commented:
@mrc0mmand

Having `kernel-headers` dependency is essential because otherwise `libbpf` lib would be built with the headers presently installed on the platform and they can be outdated. This can cause e.g. having an outdated set of program|map|attach types.
See:
```
$ cd $HOME/libbpf
$ find . -name *.[ch] | xargs grep "<linux/bpf.h>"
./src/bpf.c:#include <linux/bpf.h>
./src/bpf.h:#include <linux/bpf.h>
./src/bpf_prog_linfo.c:#include <linux/bpf.h>
./src/libbpf.c:#include <linux/bpf.h>
./src/libbpf.h:#include <linux/bpf.h>
./src/netlink.c:#include <linux/bpf.h>
./include/linux/filter.h:#include <linux/bpf.h>
```
Bumping to `0.0.6` is a bit early, the present release version for Rawhide is `0.0.5`, see [[1]](https://apps.fedoraproject.org/packages/libbpf/overview/). `0.0.6` build exists as a release candidate.

[1] https://apps.fedoraproject.org/packages/libbpf/overview/

mrc0mmand commented:
> @mrc0mmand
> 
> Having `kernel-headers` dependency is essential because otherwise `libbpf` lib would be built with the headers presently installed on the platform and they can be outdated. This can cause e.g. having an outdated set of program|map|attach types.

I see. I completely understand the need of having a dependency on `kernel-headers`, I'm more curious about the version of it, i.e. if ` 5.4.0-0.rc6.git0.1` headers have some features which are required by newer `libbpf`, thus making it unusable on CentOS 7 (which has `3.10.0-1062.9.1`).

> 
> Bumping to `0.0.6` is a bit early, the present release version for Rawhide is `0.0.5`, see [[1]](https://apps.fedoraproject.org/packages/libbpf/overview/). `0.0.6` build exists as a release candidate.

Ah, I see, will wait a bit then!

> 
> [1] https://apps.fedoraproject.org/packages/libbpf/overview/

poettering commented on src/core/cgroup.c:
```diff
@@ -2421,6 +2426,12 @@ void unit_release_cgroup(Unit *u) {
+static void unit_prune_cgroup_bpf(Unit *u) {
+#if HAVE_LIBBPF
+        u->bpf_device_control_installed = bpf_program_unref(u->bpf_device_control_installed);
+#endif
```
such ifdeffery should be left to the implementation of bpf_program_unref(), not the users

poettering commented on src/core/bpf-devices-mock.c:
```diff
@@ -0,0 +1,39 @@
+int bpf_devices_whitelist_static(BPFProgram *prog, const char *path) {
+        log_error("BPF is not supported. Mock is called");
+        return -1;
+}
```
we usually keep such "not implemented" stubs as inline functions in the headers, so that they are optimized away...

poettering commented on src/core/bpf-devices-mock.c:
```diff
@@ -0,0 +1,39 @@
+
+int bpf_devices_cgroup_init(BPFProgram **ret, CGroupDevicePolicy policy, bool whitelist) {
+        log_error("BPF is not supported. Mock is called");
+        return -1;
```
or actually:

```c
return log_error_errno(EOPNOTSUPP, "BPF …");
```

poettering commented on src/core/load-fragment.c:
```diff
@@ -4575,6 +4575,11 @@ int config_parse_ip_filter_bpf_progs(
         assert(paths);
 
+#if !(HAVE_LIBBPF)
+        log_syntax(unit, LOG_ERR, filename, line, -1, "libbpf is not found, ignoring %m");
```
do not make up error codes... i.e. no -1 please. use 0 here...
also why have this ifdeffery at all here? when bpf is disabled we should still compile the parsers for this, and this function at its end already covers the non-bpf case...

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -381,7 +382,7 @@ static int bpf_firewall_add_access_items(
                         memcpy(key_ipv4->data, &a->address, sizeof(uint32_t));
 
-                        r = bpf_map_update_element(ipv4_map_fd, key_ipv4, &value);
+                        r = bpf_map_update_elem(ipv4_map_fd, key_ipv4, &value, BPF_ANY);
```
how does libbpf do error handling? we use negative `return -errno` style returns, and what does libbpf do? if they return -1 and set errno, this needs more love

wat-ze-hex commented:
@poettering
 
Thanks for reviewing!

> Hmm, what does libbpf bring to the table for this btw? it's just a thin, trivial wrapper around the syscall, no? i.e. if it's that simple is the extra dependency really worth it? is there anything non-trivial it provides us with?

In this PR it's a syscall replacement: I'm not comfortable with introducing more changes in a single pull request, prefer smaller atomic changes. Though I can continue adding commits with functional changes in this PR, namely replacing macro assembly with C code. That's merely a matter of code organization. If only there were a way to stack PRs, so major features like this are not merged at once.
So that's up to you, I can continue growing this PR plus addressing the comments you've posted, re-requesting code reviews as more functional changes are added.

> If i understand correctly its not even a shared lib yet, i.e. unstable API?

It's a shared lib. There is [Fedora package](https://apps.fedoraproject.org/packages/libbpf) also @mrc0mmand created a [package for CentOS](https://copr.fedorainfracloud.org/coprs/mrc0mmand/systemd-centos-ci/) and `libbpf` is linked dynamically given a package.

poettering commented:
Hmm, so I think I'd prefer sticking with our syscall wrappers for now, i.e. use libbpf only if it brings more to the table for us than just syscall wrappers. I hope that makes sense?

wat-ze-hex commented:
@poettering

Sure, I'll extend this PR beyond syscall wrappers, namely replace raw bpf instructions with programs in C. Will request re-review once it's ready.

mrc0mmand commented:
Something's off:

```
 archlinux_systemd_ci: FAILED: src/shared/libsystemd-shared-244.so 
    archlinux_systemd_ci: cc  -o src/shared/libsystemd-shared-244.so 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev.c.o' 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev-device.c.o' 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev-enumerate.c.o' 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev-hwdb.c.o' 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev-list.c.o' 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev-monitor.c.o' 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev-queue.c.o' 'src/shared/5afaae1@@systemd-shared-244@sha/.._libudev_libudev-util.c.o' -fsanitize=address,undefined -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libsystemd-shared-244.so -Wl,--whole-archive src/shared/libsystemd-shared-244.a src/basic/libbasic.a src/basic/libbasic-gcrypt.a src/libsystemd/libsystemd_static.a src/journal/libjournal-client.a -Wl,--no-whole-archive -Wl,-z,relro -Wl,-z,now -fstack-protector -Wl,--gc-sections -shared -fPIC -Wl,--version-script=/build/src/shared/libshared.sym -pthread -lacl /usr/lib/libblkid.so /usr/lib64/libbpf.a /usr/lib64/libelf.a /usr/lib64/libz.a /lib64/libcap.so -lcrypt /usr/lib/libcryptsetup.so -lgcrypt /usr/lib/libidn2.so /usr/lib/libip4tc.so /usr/lib/libip6tc.so /usr/lib/libkmod.so /usr/lib/liblz4.so /usr/lib/libmount.so /usr/lib/libssl.so /usr/lib/libcrypto.so /usr/lib/libp11-kit.so -lpam -lrt /usr/lib/libseccomp.so /usr/lib/liblzma.so -lacl -lcrypt -lgcrypt -lpam -lm -lgcrypt -Wl,--end-group -Wl,-rpath,/lib64 -Wl,-rpath-link,/lib64
    archlinux_systemd_ci: /usr/bin/ld: /usr/lib64/libelf.a(elf_error.o): relocation R_X86_64_TPOFF32 against `global_error' can not be used when making a shared object; recompile with -fPIC
    archlinux_systemd_ci: /usr/bin/ld: /usr/lib64/libbpf.a(libbpf.o): relocation R_X86_64_PC32 against symbol `stderr@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
    archlinux_systemd_ci: /usr/bin/ld: final link failed: bad value
    archlinux_systemd_ci: collect2: error: ld returned 1 exit status
```
